### ISSUE TITLE
[filter-urls][escape-filenames] Decide which file extensions to download by default

### ISSUE THREAD

**Issue by anishagarwal2502** [coursera-dl/coursera-dl#443]
```
Downloading: basic-statistics\06_sampling-distributions\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\03_5-05-three-distributions.mp4
Resume downloading https://d3c33hcgiwev3.cloudfront.net/ssagNqshEeWhDQqBYtXuiQ.processed/full/540p/index.mp4?Expires=1455580800&Signature=OVFQPzZJcnMjyteID-poeKwWeMRN9SwjpYs5TIAnmPatclBeq~x0GFUmWMuy-vhB3ngEqn7ElBkuA1ikEtAKsHXTHxFBikqi-u4qer5h1ZdDH1vJ1pO1xQ414h-~6b0B~hIAy8scHsIibiKtbnE4RGuM5uBiDsrucEL9mE3upRA_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A -> basic-statistics\06_sampling-distributions\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\03_5-05-three-distributions.mp4
Starting new HTTPS connection (77): d3c33hcgiwev3.cloudfront.net
basic-statistics\06_sampling-distributions\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\03_5-05-three-distributions.mp4 already downloaded
Downloading: basic-statistics\06_sampling-distributions\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\04_reference_Naked_Statistics_Stripping_the_Dread_fro.html?id=bgfjfc_crtac&redir_esc=y
Downloading https://books.google.nl/books/about/Naked_Statistics_Stripping_the_Dread_fro.html?id=BgFJfC_CrTAC&redir_esc=y -> basic-statistics\06_sampling-distributions\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\04_reference_Naked_Statistics_Stripping_the_Dread_fro.html?id=bgfjfc_crtac&redir_esc=y
Starting new HTTPS connection (1): books.google.nl
Traceback (most recent call last):
  File "E:\coursera\\coursera-dl", line 6, in <module>
    coursera_dl.main()
  File "E:\coursera\coursera\coursera_dl.py", line 1056, in main
    if download_class(args, class_name):
  File "E:\coursera\coursera\coursera_dl.py", line 1034, in download_class
    return download_on_demand_class(args, class_name)
  File "E:\coursera\coursera\coursera_dl.py", line 1015, in download_on_demand_class
    args.resume
  File "E:\coursera\coursera\coursera_dl.py", line 528, in download_lectures
    downloader.download(url, lecfn, resume=resume)
  File "E:\coursera\coursera\downloaders.py", line 50, in download
    self._start_download(url, filename, resume)
  File "E:\coursera\coursera\downloaders.py", line 366, in _start_download
    f = open(filename, 'ab') if resume else open(filename, 'wb')
IOError: [Errno 22] invalid mode ('wb') or filename: u'basic-statistics\\06_sampling-distributions\\02_sampling-distribution-of-sample-mean-and-central-limit-theorem\\04_reference_Naked_Statistics_Stripping_the_Dread_fro.html?id=bgfjfc_crtac&redir_esc=y'
```

`E:\coursera>coursera-dl -u u****** -p p***** --resume --on-demand basic-statistics`

![image](https://cloud.githubusercontent.com/assets/6421582/13035477/5e62abfc-d377-11e5-84ab-dbb989284f3f.png)

Link  is :  https://books.google.nl/books/about/Naked_Statistics_Stripping_the_Dread_fro.html?id=BgFJfC_CrTAC&redir_esc=y

**Comment by balta2ar** [comment]
@anishagarwal2502 Thank you for the report! As a temporary measure try adding this argument: `--formats "pdf mp4 srt"` (note the double quotes). Name other file formats that you need to download so that this buggy link is excluded.

**Comment by anishagarwal2502** [comment]
Thanks and Your Welcome for such good work!

**Comment by balta2ar** [comment]
Speaking of formats and links to google books. You see, there is a slight problem with new asset extractor. It literally grabs every link and downloads it. Most of the time it's the right thing to do, but occasionally you shouldn't do that.

I've scanned through 56 courses and extracted `modules` variable from each of them. So basically I grabbed extensions, names and URLs in the way they are fed to the downloader. Here is the result of `cut -f1 ext.txt | sort | uniq -c | sort -rn -k1`:
```
   2907 mp4
   2246 en.txt
   2246 en.srt
    912 pdf
    777 ru.txt
    777 ru.srt
    262 es.txt
    262 es.srt
    220 zh-CN.txt
    220 zh-CN.srt
    184 vi.txt
    184 vi.srt
    151 ko.txt
    151 ko.srt
    141 html
    117 pt-BR.txt
    117 pt-BR.srt
    102 zip
    100 pt-PT.txt
    100 pt-PT.srt
     93 uk.txt
     93 uk.srt
     68 nl.txt
     68 nl.srt
     60 ar.txt
     60 ar.srt
     59 it.txt
     59 it.srt
     57 fr.txt
     57 fr.srt
     53 el.txt
     53 el.srt
     47 ipynb
     45 de.txt
     45 de.srt
     41 tr.txt
     41 tr.srt
     40 my.txt
     40 my.srt
     40 fa.txt
     40 fa.srt
     39 csv
     36 sr.txt
     36 sr.srt
     32 ro.txt
     32 ro.srt
     32 pptx
     32 id.txt
     32 id.srt
     31 tl.txt
     31 tl.srt
     29 php
     28 xlsx
     27 mr.txt
     27 mr.srt
     27 he.txt
     27 he.srt
     27 bg.txt
     27 bg.srt
     26 no.txt
     26 no.srt
     21 pl.txt
     21 pl.srt
     21 cs.txt
     21 cs.srt
     20 be
     18 hr.txt
     18 hr.srt
     17 hu.txt
     17 hu.srt
     17 htm
     16 lt.txt
     16 lt.srt
     15 com
     13 py
     12 sq.txt
     12 sq.srt
     10 xls
     10 et.txt
     10 et.srt
      9 org
      8 sas
      8 mn.txt
      8 mn.srt
      8 aspx
      7 ja.txt
      7 ja.srt
      7 fi.txt
      7 fi.srt
      6 txt
      6 png
      6 ms.txt
      6 ms.srt
      6 hi.txt
      6 hi.srt
      6 com)
      6 
      5 sv.txt
      5 sv.srt
      5 sl.txt
      5 sl.srt
      5 java
      4 ru
      4 lv.txt
      4 lv.srt
      4 js
      4 fas
      4 docx
      3 pt.txt
      3 pt.srt
      3 pdf?response-content-type=application%2foctet-stream&a=1&response-content-disposition=attachment
      3 html 
      3 epub?response-content-type=application%2foctet-stream&a=1&response-content-disposition=attachment
      3 edu
      2 tsv
      2 th.txt
      2 th.srt
      2 pdf 
      2 ods
      2 mov
      2 json
      2 jp
      2 html?_r=0
      2 data
      1 zip?response-content-type=application%2foctet-stream&a=1&response-content-disposition=attachment
      1 ypavibta_t84v0odchhczzbknfp3qjqa4sauqogv0igsqb85ma2i_ip5eljbfv0dcchz1nzmteer7o9t9jpj079-vdf948nskios6rph06s2cvqahkthk0cbjnbapj76ubfmq503rfkbw48-redkuhmm7xwjk-j-y4o247uauuskyqnz4t3yscrlyuhpjotxqcmbylynck-b85yzs0xubxj4xwctopewz3sek37hcons6xlogbt5pegmdc3x_9-mjio4g-af2qjjsy9ziwcd0d_nhtqgbzyghmbqom-ugrirghn01hjwxetlxumfrkkiiyxasmke6gpdsutzpzqcttwcsrfpvdtj06gyhxjmp0ljgijypk7ldqf3h3f6ilta
      1 xml
      1 ur.txt
      1 ur.srt
      1 u4f02fpu8cveioeaveh3pytdntarjbyzyhigjfodbxsywgvqb0rd-dhdgbcygy3zs-tn62p7m08wdsmxoq7n1vm-hxv5uf4gft-i8n_pbbvfqe4bjjnlqj5eek8cuxbhqc8pesyaeijrgmyf3lxvzhpq-ut2ipxkzgr1bxwtpzieja8wnnfsxkz46fhpt1cuqqxktkhcolibotm2j5derfgsluulj92-hui4hwyjt6hat5hyl5w6dzzv2aifotsbp_l9kcefdg51fpkyyxradyao4gbwdxrau1iwz1cbujan4mmtmsjfsg_2vkidpohjxzln_cbt2c9mfmvgzjoq7q
      1 transform
      1 svg
      1 staged_decision_function
      1 sk.txt
      1 sk.srt
      1 shtml 
      1 scope
      1 ru 
      1 reonc213kcbod42ekdo1vk8pd59qwmne0xc_tqeuvmnxffnibe3pqjemnvlgmylrv22tkhddx2-8fhpyhaqxrbz-4op4gz5kywpk347kny14avlhjonnmtckqkvfgsscmimydiwosvcmfnmte6tuinu5hzoxxx6-cpptjkoaezuvnuv8nhoizfg1ui77g2u-q2jzp6w3pzdt4q4grbmignvzzqmtvlp_ltbcbromgjmsetg4qdcbcaewczpxozpzvv6ut4hytiqk3u9elnyk5unvl7qpn7dgtnfarzvxqhbnng7jhycj1ojxa5ts8psj
      1 rdata
      1 pl?lang=ru&blang=ru&page=book&id=42011
      1 php?op=bio
      1 pdf?accepttc=true&seq=1#page_scan_tab_contents
      1 pdf?accepttc=true
      1 pdf+html
      1 names
      1 mk.txt
      1 mk.srt
      1 md#use-sparks-filter-method-to-return-vertices-in-the-graph
      1 md#start-the-cloudera-vm-and-upload-the-datasets-to-hdfs
      1 md#run-the-spark-shell
      1 md#join-two-datasets-with-outerjoinvertices
      1 md#join-two-datasets-with-joinvertices
      1 md#import-the-vertices-1
      1 md#import-the-vertices
      1 md#import-the-graphx-libraries
      1 md#import-the-graphstream-library
      1 md#import-the-breezeviz-library
      1 md#graph-the-results
      1 md#define-a-min-and-max-function-for-sparks-reduce-method
      1 md#define-a-function-to-calculate-the-degree-histogram
      1 md#create-a-new-return-type-for-the-joined-vertices
      1 md#create-a-new-graph-by-adding-the-continents-dataset
      1 md#create-a-new-dataset
      1 md#create-a-graph
      1 md#count-the-number-of-vertices-and-edges
      1 md#compute-the-histogram-data-of-the-degree-of-connectedness
      1 md#compute-min-and-max-degrees
      1 md#calculate-the-probability-distribution-for-the-degree-histogram
      1 md
      1 ly
      1 krxvy5mgovis
      1 kk.txt
      1 kk.srt
      1 k-dizzue9rke
      1 jsp
      1 jpg
      1 io#contributing
      1 ibooks?response-content-type=application%2foctet-stream&a=1&response-content-disposition=attachment
      1 html?ref=science
      1 html?iid=article_sidebar
      1 html?ca=drs-
      1 html?_r=2
      1 html?_r=1
      1 html)
      1 html%0d%0abrowser:%20chrome%2045%0d%0aoperating%20system:%20mac%20os%20x%0d%0a%0d%0aerror%20code:%20otgumjcuntgumtc3
      1 html#dictionaries
      1 html#conferences_2007
      1 html#
      1 htm?utm_source=nl&utm_medium=email&utm_campaign=09sep14
      1 help_hc_en-2dus_categories_200144339-2dsignature-2dtrack-2dverified-2dcertificates&d=bqmfaq&c=8huwfzcy2z-za5rbplktoq&r=ni8olixhte1fnvx0frxhyf4uxczjz052sbz7j2bq8ou&m=aox2na1-0qlqwdi2ufxdcnoyuzvhpp4f1zmpvfpenn8&s=-gje3eqbaeevrm_c9vjdnriahl1jvbtpgpv-v99dl6e&e=
      1 help_hc_en-2dus_articles_207033333-2dspecialization-2dpayments&d=bqmfaq&c=8huwfzcy2z-za5rbplktoq&r=ni8olixhte1fnvx0frxhyf4uxczjz052sbz7j2bq8ou&m=aox2na1-0qlqwdi2ufxdcnoyuzvhpp4f1zmpvfpenn8&s=apswd779hrb3dihyi1lnqwmzd89hb9fwqqsjec8f4ng&e=
      1 help_hc_en-2dus_articles_201523175-2dfinancial-2daid&d=bqmfaq&c=8huwfzcy2z-za5rbplktoq&r=ni8olixhte1fnvx0frxhyf4uxczjz052sbz7j2bq8ou&m=aox2na1-0qlqwdi2ufxdcnoyuzvhpp4f1zmpvfpenn8&s=i8kuqskashld5brhwuxa_pxnc6v6m0yawf2uc_tzqbw&e=
      1 groupby
      1 git
      1 full?sid=8ff4f86b-de05-4697-9ec8-e20b337588da
      1 full
      1 docx 
      1 doc
      1 da.txt
      1 da.srt
      1 com%2fsite%2fhowthebrainlearn%2f&sa=d&sntz=1&usg=afqjcnf29qp6kqt2oanrwpmlnoxh9mmsda
      1 com%2flhtl_mooc&sa=d&sntz=1&usg=afqjcngxi5geixukxf8avdmsryotsznmaa
      1 co
      1 cfm?id=963772
      1 bn.txt
      1 bn.srt
      1 ba_ipython_notebook
      1 aspx?utm_source=coursera&utm_medium=incoursemessaging&utm_campaign=career%20success
      1 aspx?id=2800
      1 arduinorc
      1 am
      1 al
      1 action
      1 a7gwem4kswlodv4nzkhx75dkv--4kwjbcdxw-1xu1mvh_a6wt8txp1gvmisxfrjfzh-kfc_aubcfjof1t3qig5e_h6araj4yrlm9rnbpeohnu54nnvrtotbi6c49qzkfc4lgf7bktpple1-hal8jjly26vavlfablsi4ae8arwffaldhwpzzfhynbaomcwf976gf0rihpf2nm3zs1lyxvauwe5kopstsstpwb_vcqd_qwpfxbhm6nvv-obxuptekqx9x5ui9zh2w-ywh8uls5tn8f3f5h5uc9som5gyatthfnydkytscongi0kdwnt58iyvsgatbodbqkhyb8ssao8tcncjzz5udz6onmdmqe1e
      1 3636&rep=rep1&type=pdf
      1 0
      1  

```

As you can see, most of the time file extensions are fine, but sometimes we see extensions like "jp", "com" and even "md#use-sparks-filter-method-to-return-vertices-in-the-graph".

@rbrito Any thoughts on how to deal with that reliably? Maybe we could have a heuristic to determine whether the extension is "bad" or "good". Though to my opinion, that smells.

**Comment by balta2ar** [comment]
@rbrito ping

**Comment by rbrito** [comment]
Hi, everybody. Some preliminary thoughts:

* I think that we could/should try to clean up the URLs that we grab to strip everything after a `#` sign...
* Also, downloads from external sites are likely to fail, especially if they point to the root of the site (like, say,  https://www.google.com/ or something like that).

I also noticed (after thinking for a while on one bug reported a few weeks ago) that we are not using the function to clean filenames for "new style" courses... In that function, I implemented URL unescaping... Perhaps it would be useful to call that while grabbing the URLs/filenames?

Thanks,

Rogério.
